pkg/services/dashboards/database/database.go

```diff
@@ -962,17 +965,236 @@ func (d *dashboardStore) FindDashboards(ctx context.Context, query *dashboards.F
+type SQLBuilder struct {
+	Params []any
+}
+
+func (sb *SQLBuilder) Arg(v any) (sql string) { sb.Params = append(sb.Params, v); return "?" }
+
+var searchSQL = template.Must(template.New("").Parse(`
+{{- /*
+####
+#### Here come different implementations for the innermost full-text search query.
+#### A query is expected to return a list of entities with their (OrgID, UID, Kind, Title, ParentFolderUID)
+####
+*/ -}}
+{{- define "text-search" -}}
+SELECT org_id, 'dashboard' as kind, uid, id, folder_uid as parent_uid, title FROM dashboard WHERE title {{ .Dialect.LikeStr }} {{ .Arg .Text }} AND NOT is_folder
+UNION ALL
+SELECT org_id, 'folder' as kind, uid, 0, parent_uid, title FROM folder WHERE title {{ .Dialect.LikeStr }} {{ .Arg .Text }}
+{{ end }}
+{{- define "text-search-mysql-fts" -}}
+SELECT org_id, 'dashboard' as kind, uid, id, folder_uid as parent_uid, title FROM dashboard WHERE MATCH(title) AGAINST({{ .Arg .Text }}) AND NOT is_folder
+UNION ALL
+SELECT org_id, 'folder' as kind, uid, 0, parent_uid, title FROM folder WHERE MATCH(title) AGAINST({{ .Arg .Text }})
+{{ end }}
+{{- define "text-search-legacy" -}}
+SELECT org_id, 'dashboard' as kind, uid, id, folder_uid as parent_uid, title FROM dashboard WHERE title {{ .Dialect.LikeStr }} {{ .Arg .Text }} AND NOT is_folder
+UNION ALL
+SELECT org_id, 'folder' as kind, uid, id, folder_uid as parent_uid, title FROM dashboard WHERE title {{ .Dialect.LikeStr }} {{ .Arg .Text }} AND is_folder
+{{ end }}
+
+SELECT org_id, kind, uid, entity.id, title, dashboard_tag.term as term, folder_uid, folder_title FROM (
+	SELECT DISTINCT
+		entity.org_id, entity.kind, entity.uid, entity.id, entity.title, f1.uid as folder_uid, f1.title as folder_title FROM (
+		{{- template "text-search-legacy" . -}}
+	) AS entity
+	LEFT JOIN folder f1 ON (entity.parent_uid = f1.uid AND entity.org_id = f1.org_id)
+	{{ if ne .UserID 0 }}   {{- /* If a user is not admin: need to filter by permissions */ -}}
+	LEFT JOIN folder f2 ON (f1.parent_uid     = f2.uid AND f1.org_id     = f2.org_id)
+	LEFT JOIN folder f3 ON (f2.parent_uid     = f3.uid AND f2.org_id     = f3.org_id)
+	LEFT JOIN folder f4 ON (f3.parent_uid     = f4.uid AND f3.org_id     = f4.org_id)
+	INNER JOIN permission p ON 
+	p.role_id IN (
+		SELECT ur.role_id FROM user_role AS ur    WHERE ur.user_id = {{ .Arg .UserID }} AND (ur.org_id = {{ .Arg .OrgID }} OR ur.org_id = 0)
+		{{ if .TeamIDs }}
+		UNION
+		SELECT tr.role_id FROM team_role AS tr    WHERE tr.team_id IN (
+			{{- range $i, $el := .TeamIDs -}}
+			{{- if $i -}},{{- end -}}
+			{{- $.Arg $el -}}
+			{{- end -}}
+		) AND tr.org_id = {{ .Arg .OrgID }}
+		{{ end }}
+		UNION
+		SELECT br.role_id FROM builtin_role AS br WHERE br.role IN ({{ .Arg .Role }}) AND (br.org_id = {{ .Arg .OrgID }} OR br.org_id = 0)
+	) AND (
+		{{- /* This permission check relies on the optimised (indexed) split scope columns in the permission table */ -}}
+		(
+			p.kind = 'dashboards' AND
+			p.action = 'dashboards:read' AND
```

Comment:
nit can we use the constant [ActionDashboardsRead](https://github.com/grafana/grafana/blob/0d1462cbbbe1e4712170d946818c2299bbb0e8b1/pkg/services/dashboards/accesscontrol.go#L26C2-L26C22)?